js/modules/k6/ws/ws.go

```diff
@@ -0,0 +1,385 @@
+	connectionEnd := time.Now()
+	connectionDuration := stats.D(connectionEnd.Sub(start))
+
+	socket := Socket{
+		ctx:                ctx,
+		conn:               conn,
+		eventHandlers:      make(map[string][]goja.Callable),
+		pingSendTimestamps: make(map[string]time.Time),
+		scheduled:          make(chan goja.Callable),
+		done:               make(chan struct{}),
+	}
+
+	// Run the user-provided set up function
+	setupFn(goja.Undefined(), rt.ToValue(&socket))
+
+	if connErr != nil {
+		// Pass the error to the user script before exiting immediately
+		socket.handleEvent("error", rt.ToValue(connErr))
+		return response
+	}
+	defer conn.Close()
+
+	tags["status"] = strconv.Itoa(response.StatusCode)
+	tags["subprotocol"] = response.Header.Get("Sec-WebSocket-Protocol")
+
+	// The connection is now open, emit the event
+	socket.handleEvent("open")
+
+	// Pass ping/pong events through the main control loop
+	pingChan := make(chan string)
+	pongChan := make(chan string)
+	conn.SetPingHandler(func(msg string) error { pingChan <- msg; return nil })
+	conn.SetPongHandler(func(pingID string) error { pongChan <- pingID; return nil })
+
+	readDataChan := make(chan []byte)
+	readErrChan := make(chan error)
+
+	// Wraps a couple of channels around conn.ReadMessage
+	go readPump(conn, readDataChan, readErrChan)
+
+	// This is the main control loop. All JS code (including error handlers)
+	// should only be executed by this thread to avoid race conditions
+	for {
+		select {
+		case <-pingChan:
+			// Handle pings received from the server
+			socket.handleEvent("ping")
+
+		case pingID := <-pongChan:
+			// Handle pong responses to our pings
+			socket.trackPong(pingID)
+			socket.handleEvent("pong")
+
+		case readData := <-readDataChan:
+			socket.msgsReceived++
+			socket.handleEvent("message", rt.ToValue(string(readData)))
+
+		case readErr := <-readErrChan:
+			socket.handleEvent("error", rt.ToValue(readErr))
+
+		case scheduledFn := <-socket.scheduled:
+			scheduledFn(goja.Undefined())
+
+		case <-ctx.Done():
+			// This means that K6 is shutting down (e.g., during an interrupt)
+			socket.handleEvent("close", rt.ToValue("Interrupt"))
+			socket.closeConnection(websocket.CloseGoingAway)
+
+		case <-socket.done:
+			// This is the final exit point normally triggered by closeConnection
+			end := time.Now()
+			sessionDuration := stats.D(end.Sub(start))
+
+			samples := []stats.Sample{
+				{Metric: metrics.WSSessions, Time: end, Tags: tags, Value: 1},
+				{Metric: metrics.WSMessagesSent, Time: end, Tags: tags, Value: socket.msgsReceived},
+				{Metric: metrics.WSMessagesReceived, Time: end, Tags: tags, Value: socket.msgsSent},
+				{Metric: metrics.WSConnecting, Time: end, Tags: tags, Value: connectionDuration},
+				{Metric: metrics.WSSessionDuration, Time: end, Tags: tags, Value: sessionDuration},
+			}
+
+			for _, pingDelta := range socket.pingDeltas {
+				samples = append(samples, stats.Sample{
+					Metric: metrics.WSPing,
+					Time:   end,
```

Comment:
On a graph, this will put a clump of dots in one place, rather than a nice, repeating pattern. Which is not only aestetically pleasing, but can also reveal spikes at appropriate times that can be correlated. I'd recommend keeping an internal buffer of samples that you emit at the end of a session, similar to what `state.Samples` is for a VU.